File: projecteuler/0804/p804.py
import math

def check(x, y, n):
    return x * x + x * y + 41 * y * y <= n

N = 10 ** 16
L = int(math.pow(4 * N  / 163, 0.5))
total = 0
for y in range(-L, L + 1):
    s = int(math.pow(4 * N - 163 * y * y, 0.5))
    l = (-s - y) // 2
    r = (s - y) // 2
    total += r - l + 1
    if not check(l, y, N):
        total -= 1
    if not check(r, y, N):
        total -= 1
print(total - 1)

File: projecteuler/0804/p804.go
package main

import (
	"fmt"
	"math"
)

var N = int(math.Pow10(16))

func check(x int, y int) bool {
	return x * x + x * y + 41 * y * y <= N
}

func main() {
	L := int(math.Sqrt(4 * float64(N) / 163))
	total := 0
	for y := -L; y <= L; y++ {
		s := int(math.Sqrt(4 * float64(N) - 163 * float64(y * y)))
		l := int(-s - y) / 2
		r := int(s - y) / 2
		total += r - l + 1
		if !check(l, y) { total-- }
		if !check(r, y) { total-- }
	}
	fmt.Println(total - 1)
}
